[driver] Introduce `clang-cache-build-session`

akyrtzi opened the pull request:
`clang-cache-build-session` can be used in combination with `clang-cache`
to start a build command that indicates a "build session". After configuring
to use `clang-cache` as the compiler launcher, you can use it like this:

```
% $TOOLCHAIN/usr/bin/clang-cache-build-session ninja
```

`clang-cache-build-session` accepts a command that triggers a build and,
while the command is running, all the clang invocations that are part of
that build will be sharing the same dependency scanning daemon.

This is accomplished by having `clang-cache-build-session` set
`CLANG_CACHE_BUILD_SESSION_ID=<pid>` as environment variable which is then
passed on to the new "depscan" flag `-fdepscan-share-identifier`.

akyrtzi committed:
[driver] Introduce `clang-cache-build-session`

`clang-cache-build-session` can be used in combination with `clang-cache`
to start a build command that indicates a "build session". After configuring
to use `clang-cache` as the compiler launcher, you can use it like this:

```
% $TOOLCHAIN/usr/bin/clang-cache-build-session ninja
```

`clang-cache-build-session` accepts a command that triggers a build and,
while the command is running, all the clang invocations that are part of
that build will be sharing the same dependency scanning daemon.

This is accomplished by having `clang-cache-build-session` set
`CLANG_CACHE_BUILD_SESSION_ID=<pid>` as environment variable which is then
passed on to the new "depscan" flag `-fdepscan-share-identifier`.

akyrtzi commented:
@swift-ci Please test

dexonsmith commented:
I wonder if it'd be more robust in some way to manually start a daemon and shut it down after. On the other hand, that'd be wasteful in a null build.

dexonsmith commented on clang/tools/driver/CacheLauncherMode.cpp:
```diff
@@ -62,13 +62,16 @@ clang::handleClangCacheInvocation(SmallVectorImpl<const char *> &Args,
+      Args.append(
+          {"-fdepscan=daemon", "-fdepscan-share-identifier", SessionId});
+    } else {
+      Args.push_back("-fdepscan=inline");
```
Not really a change from this commit, but I'm curious why you're passing `=inline`, vs. `-fdepscan` or `-fdepscan=auto` (and letting the compiler figure out the best thing)?

akyrtzi replied:
Switched to passing "-fdepscan" for the non-session case.

dexonsmith replied:
FTR, @akyrtzi and I chatted offline.
- Currently, `-fdepscan=daemon` is only useful if sharing scanning state. As a result, `-fdepscan=auto` is the same as `-fdepscan=inline`.
- There are FIXMEs for `-fdepscan=daemon` to use the same daemon even when NOT sharing scanning state. This would not be a semantic change: idea is to share an OnDiskCAS connection, but to have different scanning arenas in the process, and only share `CachingOnDiskFileSystem` and `DependencyScanningService` in the same places it would have been previously. Once that's done, it'd be nice to change `-fdepscan=auto` (the default) to mean `-fdepscan=daemon` if that turns out to be faster.
- Later, the daemon could grow to do other work (not just depscanning), as a central build server.

dexonsmith reviewed (approved):
Despite my comments, this looks like a good step forward! LGTM.

akyrtzi commented:
> I wonder if it'd be more robust in some way to manually start a daemon and shut it down after. On the other hand, that'd be wasteful in a null build.

Initially I thought that's the way to go but then it occurred to me that the build may have mixed toolchains, in which case you want to dynamically run multiple daemons based on the clang versions, not share only one.

dexonsmith commented:
> > I wonder if it'd be more robust in some way to manually start a daemon and shut it down after. On the other hand, that'd be wasteful in a null build.
> 
> Initially I thought that's the way to go but then it occurred to me that the build may have mixed toolchains, in which case you want to dynamically run multiple daemons based on the clang versions, not share only one.

That makes sense. This could happen, e.g., in a bootstrap build! I suggest adding a comment to document why.

akyrtzi commented:
Updates:
* switched to passing "-fdepscan" for the non-session case.
* added a comment explaining why it is preferable to still launch daemons on-demand during a "build session".
* added test case in `test/CAS/driver-cache-launcher.c` that uses `clang-cache-build-session`.

akyrtzi commented:
@swift-ci Please test
